### PR TITLE
services/horizon/internal: Add fee_charged and max_fee to /fee_stats endpoint

### LINKED ISSUE #1372: Display fees charged in /fee_stats endpoint
In #1358 we introduced new fields in Transaction resource:
* `fee_charged` - a fee actually paid for a transaction (`txresult.FeeCharged`)
* `max_fee` - a maximum fee source account is willing to pay (`transaction.Fee`)

The `/fee_stats` endpoint displays stats for fee bids (using `max_fee` value) instead of the fees that were actually paid for transactions. This can be confusing for users that can bid even higher fees.

Basically the main problem is that field names (`accepted_fee` suffix) are really confusing in a new context.

### PR DESCRIPTION
**Pull request by abuiles**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Display fees charged and max fee in /`fee_stats` endpoint. Fixes #1372 

### Why
We have a way now to display not only `max_fee` but also `fee_charged` data on the `/fee_stats` end-point. This PR adds  info for both fields and at the same time schedules the old field for deprecation which had a confusing name as described https://github.com/stellar/go/issues/1372#issuecomment-543305613

### Known limitations

[TODO or N/A]

### CODE REVIEW

**Comment by abuiles** [comment]
@bartekn this cherry-picks on the work done in this PR https://github.com/stellar/go/pull/1953 -- you can ignore those changes for now.

File: services/horizon/internal/actions_operation_fee_stats_test.go [hunk 1]
```diff
@@ -134,29 +217,59 @@ func TestOperationFeeTestsActions_ShowMultiOp(t *testing.T) {
...
+		ht.Assert.Equal(200, result.MaxFee.P40, "p40")
+		ht.Assert.Equal(200, result.MaxFee.P50, "p50")
+		ht.Assert.Equal(200, result.MaxFee.P60, "p60")
+		ht.Assert.Equal(200, result.MaxFee.P70, "p70")
+		ht.Assert.Equal(200, result.MaxFee.P80, "p80")
+		ht.Assert.Equal(200, result.MaxFee.P90, "p90")
+		ht.Assert.Equal(200, result.MaxFee.P95, "p95")
+		ht.Assert.Equal(200, result.MaxFee.P99, "p99")
+
+		ht.Assert.Equal(50, result.FeeCharged.Min, "min")
```
Comment (abuiles) [inline services/horizon/internal/actions_operation_fee_stats_test.go]: @bartekn can fee_charged  be lower than 50? If not, we probably should make an update on the `fee_charged` entries in this test scenario?

File: services/horizon/internal/actions_operation_fee_stats_test.go [hunk 2]
```diff
@@ -172,28 +285,57 @@ func TestOperationFeeTestsActions_NotInterpolating(t *testing.T) {
...
+		ht.Assert.Equal(400, result.MaxFee.P60, "p60")
+		ht.Assert.Equal(400, result.MaxFee.P70, "p70")
+		ht.Assert.Equal(400, result.MaxFee.P80, "p80")
+		ht.Assert.Equal(16000, result.MaxFee.P90, "p90")
+		ht.Assert.Equal(16000, result.MaxFee.P95, "p95")
+		ht.Assert.Equal(16000, result.MaxFee.P99, "p99")
+
+		ht.Assert.Equal(6, result.FeeCharged.Min, "min")
+		ht.Assert.Equal(100, result.FeeCharged.Mode, "mode")
+		ht.Assert.Equal(6, result.FeeCharged.P10, "p10")
```
Comment (abuiles) [inline services/horizon/internal/actions_operation_fee_stats_test.go]: @bartekn same comment as before, does 6 even makes sense as a fee_charged?

**Comment by abuiles** [comment]
oh, I just realized I forgot to add max, will do that first thing in the morning.

File: protocols/horizon/main.go [hunk 3]
```diff
@@ -624,25 +624,44 @@ type Metrics struct {
 	TxsubTotal             LogTotalMetric `json:"txsub.total"`
 }
 
+type feeStatsBase struct {
```
Comment (tamirms) [inline protocols/horizon/main.go]: `feeDistribution` might be a better name

Comment (abuiles) [reply protocols/horizon/main.go]: updated

File: services/horizon/internal/actions_operation_fee_stats_test.go [hunk 4]
```diff
@@ -177,23 +301,54 @@ func TestOperationFeeTestsActions_NotInterpolating(t *testing.T) {
...
+		ht.Assert.Equal(100, result.FeeCharged.P20, "fee_charged_p20")
+		ht.Assert.Equal(100, result.FeeCharged.P30, "fee_charged_p30")
+		ht.Assert.Equal(100, result.FeeCharged.P40, "fee_charged_p40")
+		ht.Assert.Equal(100, result.FeeCharged.P50, "fee_charged_p50")
+		ht.Assert.Equal(100, result.FeeCharged.P60, "fee_charged_p60")
+		ht.Assert.Equal(100, result.FeeCharged.P70, "fee_charged_p70")
+		ht.Assert.Equal(100, result.FeeCharged.P80, "fee_charged_p80")
+		ht.Assert.Equal(100, result.FeeCharged.P90, "fee_charged_p90")
+		ht.Assert.Equal(100, result.FeeCharged.P95, "fee_charged_p95")
+		ht.Assert.Equal(100, result.FeeCharged.P99, "fee_charged_p99")
```
Comment (bartekn) [inline services/horizon/internal/actions_operation_fee_stats_test.go]: Can we add a test where `FeeCharged` and `FeeMax` is actually max uint32 to catch all future "wrong type" issues?

File: services/horizon/internal/actions_operation_fee_stats_test.go (hunk 1 above)
Comment (bartekn) [reply services/horizon/internal/actions_operation_fee_stats_test.go]: Have you found an answer to the question? I think the code is outdated so can't check it out.

File: services/horizon/internal/actions_operation_fee_stats_test.go (hunk 2 above)
Comment (bartekn) [reply services/horizon/internal/actions_operation_fee_stats_test.go]: Have you found an answer to the question? I think the code is outdated so can't check it out.

File: services/horizon/internal/actions_operation_fee_stats_test.go (hunk 1 above)
Comment (abuiles) [reply services/horizon/internal/actions_operation_fee_stats_test.go]: @bartekn yes,  after scrolling around I came to the conclusion that it's done based on the data we have in the DB, which is fine then.

File: protocols/horizon/main.go [hunk 5]
```diff
@@ -624,25 +624,44 @@ type Metrics struct {
...
+	P20  int `json:"p20,string"`
+	P30  int `json:"p30,string"`
+	P40  int `json:"p40,string"`
+	P50  int `json:"p50,string"`
+	P60  int `json:"p60,string"`
+	P70  int `json:"p70,string"`
+	P80  int `json:"p80,string"`
+	P90  int `json:"p90,string"`
+	P95  int `json:"p95,string"`
+	P99  int `json:"p99,string"`
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply protocols/horizon/main.go]: fee bump is going to make this int64, so if you make a change, make it safe for int64 @bartekn

File: services/horizon/internal/operationfeestats/main.go [hunk 6]
```diff
@@ -12,22 +12,39 @@ import (
...
+	FeeP20  int64
+	FeeP30  int64
+	FeeP40  int64
+	FeeP50  int64
+	FeeP60  int64
+	FeeP70  int64
+	FeeP80  int64
+	FeeP90  int64
+	FeeP95  int64
+	FeeP99  int64
```
[the remark this answers is not in the document]
Comment (abuiles) [reply services/horizon/internal/operationfeestats/main.go]: sure,  done!

File: services/horizon/internal/actions_operation_fee_stats.go [hunk 7]
```diff
@@ -16,26 +17,35 @@ import (
 
 var _ actions.JSONer = (*FeeStatsAction)(nil)
 
+// this struct is very similar to hProtocol.feeStats but drops the usage of int
+// in favor of int64
+type feeStats struct {
```
Comment (abuiles) [inline services/horizon/internal/actions_operation_fee_stats.go]: this is a temporary workaround while we update hProtocol.FeeStats

File: services/horizon/internal/db2/history/operation.go [hunk 8]
```diff
@@ -44,19 +44,34 @@ func (r *Operation) UnmarshalDetails(dest interface{}) error {
...
+		    ceil(percentile_disc(0.10) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p10",
+		    ceil(percentile_disc(0.20) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p20",
+		    ceil(percentile_disc(0.30) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p30",
+		    ceil(percentile_disc(0.40) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p40",
+		    ceil(percentile_disc(0.50) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p50",
+		    ceil(percentile_disc(0.60) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p60",
+		    ceil(percentile_disc(0.70) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p70",
+		    ceil(percentile_disc(0.80) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p80",
+		    ceil(percentile_disc(0.90) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p90",
+		    ceil(percentile_disc(0.95) WITHIN GROUP (ORDER BY fee_charged/operation_count))::bigint AS "fee_charged_p95",
```
[the remark this answers is not in the document]
Comment (abuiles) [reply services/horizon/internal/db2/history/operation.go]: @bartekn fixed!  We are using tabs here

File: services/horizon/internal/actions_operation_fee_stats.go [hunk 9]
```diff
@@ -57,45 +43,75 @@ func (action *FeeStatsAction) JSON() error {
...
+	action.FeeStats.MaxFee.P20 = int(cur.FeeP20)
+	action.FeeStats.MaxFee.P30 = int(cur.FeeP30)
+	action.FeeStats.MaxFee.P40 = int(cur.FeeP40)
+	action.FeeStats.MaxFee.P50 = int(cur.FeeP50)
+	action.FeeStats.MaxFee.P60 = int(cur.FeeP60)
+	action.FeeStats.MaxFee.P70 = int(cur.FeeP70)
+	action.FeeStats.MaxFee.P80 = int(cur.FeeP80)
+	action.FeeStats.MaxFee.P90 = int(cur.FeeP90)
+	action.FeeStats.MaxFee.P95 = int(cur.FeeP95)
+	action.FeeStats.MaxFee.P99 = int(cur.FeeP99)
```
[the remark this answers is not in the document]
Comment (abuiles) [reply services/horizon/internal/actions_operation_fee_stats.go]: updated to int64

File: services/horizon/internal/actions_operation_fee_stats_test.go (hunk 4 above)
Comment (abuiles) [reply services/horizon/internal/actions_operation_fee_stats_test.go]: update to int64, we need to keep the `int` for old types for now, but we'll remove  them in 0.25.0

**Comment by abuiles** [comment]
@bartekn I addresses your comments, thanks!

**Comment by bartekn** [comment]
@abuiles FYI, changed `LastLedger` type to `uint32`: https://github.com/stellar/go/pull/1964/commits/1df8d94a490609255b41c5e586ac3873a9a405e2.

**Pull request merged**